### ISSUE TITLE
iree_allocator_malloc and policy on zeroing buffers

### ISSUE THREAD

**Issue by ewconnell** [google/iree#4737]
I have some questions after reviewing the `iree/bindings/tflite` implementation. In the `_TfLiteInterpreterAllocate` function `iree/bindings/tflite/interpreter.c::255` there is a call to `iree_allocator_malloc`. From what I can tell this function is the mechanism used for general memory allocation (which might be on the host or device). The implementation as documented always zeros the newly allocated buffer by specifying `IREE_ALLOCATION_MODE_ZERO_CONTENTS`
```C
IREE_API_EXPORT iree_status_t IREE_API_CALL iree_allocator_malloc(
    iree_allocator_t allocator, iree_host_size_t byte_length, void** out_ptr) {
  if (!allocator.alloc) {
    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
                            "allocator has no alloc routine");
  }
  return allocator.alloc(allocator.self, IREE_ALLOCATION_MODE_ZERO_CONTENTS,
                         byte_length, out_ptr);
}
```
It doesn't appear there is an api for allocating a non-initialized buffer. The function `iree_allocator_malloc` is correctly documented, however `malloc` in the C library does not initialize memory and perhaps will lead to confusion by users.

There are several examples in the code as shown in `_TfLiteInterpreterAllocate`, where memory is allocated and then zeroed again.
```C
static iree_status_t _TfLiteInterpreterAllocate(
    const TfLiteModel* model, TfLiteInterpreter** out_interpreter) {
  iree_host_size_t interpreter_size = _TfLiteInterpreterCalculateSize(model);
  TfLiteInterpreter* interpreter = NULL;
  IREE_RETURN_IF_ERROR(iree_allocator_malloc(model->allocator, interpreter_size,
                                             (void**)&interpreter));
  memset(interpreter, 0, interpreter_size);  // <--- zeroed a second time
  interpreter->allocator = model->allocator;
``` 
I'm guessing most developers might not read the docs and assume the memory hasn't been cleared for them.

There is also a secondary performance issue. It's desirable to create an uninitialized output buffer for operations when you know that the operation will overwrite every element, especially when the operation is memory bandwidth limited. I found this to be a significant factor in performance tests.

Thoughts?

**Comment by benvanik** [comment]
This is a TODO to clean up - ideally given the naming iree_allocator_malloc should not zero the memory and instead we should have an iree_allocator_calloc that sets the ZERO flag. Since iree_allocator_malloc zeros today there are places in the code expecting that and it needs a scrub to find them. This would be good cleanup and make the code more readable. That said, IREE works hard to not allocate in the steady state and always zeroing memory is just pure safety and no perf hit. If anything, this means iree_allocator_malloc should just be renamed to iree_allocator_calloc and we shouldn't even have the non-zeroing variant.

Now, whether to zero-initialize HAL buffers is a different thing, though, but also not really a big deal in IREE. In general we have to ensure buffers are zeroed whenever we can't know for certain that the memory is going to be initialized by the user. We can't let user code (the compiled modules we are executing) access uninitialized memory, afterall. Instead, we design around it and end up getting wins in other areas too: buffers are initialized on first acquisition from the underlying allocator and then heavily reused - and only for specific purposes. For example, memory allocated for HAL buffers will never be reused for anything but those, meaning that once we've initialized them once (to ensure we aren't leaking in memory from outside the safety boundary) we can reuse them without reinitializing for their lifetime: we don't care if a user's program can read its own buffer memory, only that it can't read executable/data pages from outside.

In other words: HAL buffers for the host-local device backend are allocated independently from runtime allocations and must always be zeroed on acquisition from the system, but that doesn't much matter as we will be heavily reusing them and the initial cost is cheap.

**Comment by ewconnell** [comment]
I see. I'm not up to speed yet on HAL buffer reuse frequency. 
Should the contents of a reused HAL buffer from a previous operation be thought of differently than uninitialized memory? It seems they would both be random.

**Comment by benvanik** [comment]
My goal this month is to write the buffer reuse code so unfortunately there's nothing I can point you at yet :)

[VMA](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/) is very similar to what we will be doing, only we'll be emitting (effectively) the code for what it does in the compiler vs. requiring all the runtime tracking. This has a good overview of some of the types of schemes we'll be using: https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/custom_memory_pools.html

The other relevant thing is the CUDA stream-ordered allocator; conceptually how we allocate in IREE is the same, only we are emitting the code to track this vs. requiring the runtime to do so: https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html#group__CUDART__MEMORY__POOLS

**Comment by benvanik** [comment]
Oh and RE reuse: it's part defense in depth and part compatibility with non-local execution models (sandbox/out-of-process, cross-device, remote/rpc, etc). Each running program by design cannot access any memory used by any other program - either another IREE compiled program or the hosting application. The two ways IREE programs can access memory is by host-side operations through the IREE VM and by device-side operations through the HAL buffer. The VM operations are easy to check and critical - not checking there leads to all kinds of attack vectors - and since the VM is just issuing commands there's no real performance cost. The HAL operations are all checked on buffer boundaries during command recording and the assurance the HAL allocator implementations provide is that buffer bytes between iree_hal_device_t instances cannot be read/written by any other iree_hal_device_t. In the CPU case this means a malloc of a buffer in device B must not see the bytes of a previous malloc/free from device A. The programs we generate hang on to the HAL buffers they allocate and reuse them (see above), so the HAL allocator only needs to ensure the initial allocation from it is zerod (or filled with 0xCC, etc). When new buffers are allocated the contract within the HAL is that two allocations from the same device _may_ not be zeroed and _may_ contain arbitrary contents, but that those contents will be results from operations performed only by that same device. This is safe as there's no way for the user program to exploit this reuse as it could just as easily have issued a read operation against the memory it's already able to read. It helps that we aren't trying to guarantee correctness of a user program and instead only that a correctly written user program executes safely. If you use-after-free or have hazards your program will malfunction but you should never be able to to escape the sandbox.

The upside of all this is that it leads to a better design, and enables multitenancy for things like a device-side execution sandbox where one out-of-process thread pool is running multiple program workloads simultaneously.
But really we zero because to not do so is not worth a single potential vulnerability :)

**Comment by ewconnell** [comment]
Ahh good point, I wasn't thinking about the security aspects. Thank you for such a detailed explanation!

**Comment by benvanik** [comment]
Thanks for asking! This stuff hasn't been documented at all and now we have at least *something* to point people too, as it's a total valid question. Please ask more as you come across things :)
